beacon-chain/core/helpers/rewards_penalties.go

```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
 	validatorBalance := EffectiveBalance(state, validatorIndex)
 	return baseReward + validatorBalance*epochsSinceFinality/params.BeaconConfig().InactivityPenaltyQuotient/2
 }
+
+// GetBalance returns the Gwei balance of a validator by its index
+//
+// Spec pseudocode definition:
+//   def get_balance(state: BeaconState, index: ValidatorIndex) -> Gwei:
+//     """
+//     Return the balance for a validator with the given ``index``.
+//     """
+//     return state.balances[index]
+func GetBalance(state *pb.BeaconState, validatorIndex uint64) uint64 {
+	return state.ValidatorBalances[validatorIndex]
+}
+
+// SetBalance sets the Gwei balance of a validator by its index
+//
+// Spec pseudocode definition:
+// def set_balance(state: BeaconState, index: ValidatorIndex, balance: Gwei) -> None:
+//     """
+//     Set the balance for a validator with the given ``index`` in both ``BeaconState``
+//     and validator's rounded balance ``high_balance``.
+//     """
+//     validator = state.validator_registry[index]
+//     HALF_INCREMENT = HIGH_BALANCE_INCREMENT // 2
+//     if validator.high_balance > balance or validator.high_balance + 3 * HALF_INCREMENT < balance:
+//         validator.high_balance = balance - balance % HIGH_BALANCE_INCREMENT
+//     state.balances[index] = balance
+func SetBalance(state *pb.BeaconState, validatorIndex uint64, balance uint64) {
+	validator := state.ValidatorRegistry[validatorIndex]
+	halfIncrement := params.BeaconConfig().HighBalanceIncrement / 2
+	if validator.HighBalance > balance || validator.HighBalance+3*halfIncrement < balance {
+		validator.HighBalance = balance - balance%params.BeaconConfig().HighBalanceIncrement
+	}
+	state.ValidatorBalances[validatorIndex] = balance
```

Comment:
I think you need to return the state at the end

Reply:
I don't know. they get pointer to state and change it. what do you think?

Reply:
@terencechain ?

Reply:
we should return the state there, similar to other helper functions. These are pure functions

Reply:
thanks